std/experimental/testing/testsuite.d

```diff
@@ -0,0 +1,291 @@
+            utWriteln(_options.testsToRun);
+            return false;
+        }
+
+        immutable elapsed = doRun();
+
+        if(!numTestsRun)
+        {
+            utWriteln("Did not run any tests!!!");
+            return false;
+        }
+
+        utWriteln("\nTime taken: ", elapsed);
+        utWrite(numTestsRun, " test(s) run, ");
+        const failuresStr = text(_failures.length, " failed");
+        if(_failures.length)
+        {
+            utWriteRed(failuresStr);
+        }
+        else
+        {
+            utWrite(failuresStr);
+        }
+
+        void printAbout(string attr)(in string msg)
+        {
+            const num = _testData.filter!(a => mixin("a. " ~ attr)).count;
+            if(num)
+            {
+                utWrite(", ");
+                utWriteYellow(num, " " ~ msg);
+            }
+        }
+
+        printAbout!"hidden"("hidden");
+        printAbout!"shouldFail"("failing as expected");
+
+        utWriteln(".\n");
+
+        if(_failures.length)
+        {
+            utWritelnRed("Unit tests failed!\n");
+            return false; //oops
+        }
+
+        utWritelnGreen("OK!\n");
+
+
+        return true;
+    }
+
+private:
+
+    const(Options) _options;
+    const(TestData)[] _testData;
+    TestCase[] _testCases;
+    string[] _failures;
+    StopWatch _stopWatch;
+
+    /**
+     * Runs the tests with the given options.
+     * Returns: how long it took to run.
+     */
+    Duration doRun()
+    {
+        auto tests = getTests();
+        _stopWatch.start();
+
+        if(_options.multiThreaded)
+        {
+            _failures = reduce!((a, b) => a ~ b)(_failures,
+                                                 taskPool.amap!runTest(tests));
+        }
+        else
+        {
+            foreach(test; tests) _failures ~= test();
+        }
+
+        handleFailures();
+
+        _stopWatch.stop();
+        return cast(Duration)_stopWatch.peek();
+    }
+
+
+    auto getTests()
+    {
+        auto tests = _testCases.dup;
+        if(_options.random)
+        {
+            import std.random;
+            auto generator = Random(_options.seed);
+            tests.randomShuffle(generator);
+            utWriteln("Running tests in random order. ",
+                      "To repeat this run, use --seed ", _options.seed);
+        }
+        return tests;
+    }
+
+    void handleFailures() const
+    {
+        if(!_failures.empty) utWriteln("");
+        foreach(failure; _failures)
+        {
+            utWrite("Test ", failure, " ");
+            utWriteRed("failed");
+            utWriteln(".");
+        }
+        if(!_failures.empty) utWriteln("");
+    }
+
+    @property ulong numTestsRun() @safe const pure
+    {
+        return _testCases.map!(a => a.numTestsRun).reduce!((a, b) => a + b);
+    }
+}
+
+/**
+ * Replace the D runtime's normal unittest block tester with our own
+ */
+shared static this()
+{
+    Runtime.moduleUnitTester = &moduleUnitTester;
```

Comment:
This might not work with phobos as shared library, because in that case all phobos constructors will be run.
Can you wrap it in some template and instantiate it in some entry point, like `runTests` or so?